pkg/api/api.go

```diff
@@ -0,0 +1,304 @@
+		a.registerRoute("/api/v1/rules/{namespace}/{groupName}", http.HandlerFunc(r.GetRuleGroup), true, "GET")
+		a.registerRoute("/api/v1/rules/{namespace}", http.HandlerFunc(r.CreateRuleGroup), true, "POST")
+		a.registerRoute("/api/v1/rules/{namespace}/{groupName}", http.HandlerFunc(r.DeleteRuleGroup), true, "DELETE")
+
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/api/v1/rules", http.HandlerFunc(r.PrometheusRules), true, "GET")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/api/v1/alerts", http.HandlerFunc(r.PrometheusAlerts), true, "GET")
+
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules", http.HandlerFunc(r.ListRules), true, "GET")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}", http.HandlerFunc(r.ListRules), true, "GET")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}/{groupName}", http.HandlerFunc(r.GetRuleGroup), true, "GET")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}/{groupName}", http.HandlerFunc(r.GetRuleGroup), true, "GET")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}", http.HandlerFunc(r.CreateRuleGroup), true, "POST")
+		a.registerRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}/{groupName}", http.HandlerFunc(r.DeleteRuleGroup), true, "DELETE")
+	}
+}
+
+// // RegisterRing registers the ring UI page associated with the distributor for writes.
+func (a *API) RegisterRing(r *ring.Ring) {
+	a.registerRoute("/ring", r, false)
+}
+
+// RegisterStoreGateway registers the ring UI page associated with the store-gateway.
+func (a *API) RegisterStoreGateway(s *storegateway.StoreGateway) {
+	a.registerRoute("/store-gateway/ring", http.HandlerFunc(s.RingHandler), false)
+}
+
+// RegisterCompactor registers the ring UI page associated with the compactor.
+func (a *API) RegisterCompactor(c *compactor.Compactor) {
+	a.registerRoute("/compactor/ring", http.HandlerFunc(c.RingHandler), false)
+}
+
+// RegisterQuerier registers the Prometheus routes supported by the
+// Cortex querier service. Currently this can not be registered simultaneously
+// with the QueryFrontend.
+func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine, distributor *distributor.Distributor) {
+	api := v1.NewAPI(
+		engine,
+		queryable,
+		querier.DummyTargetRetriever{},
+		querier.DummyAlertmanagerRetriever{},
+		func() config.Config { return config.Config{} },
+		map[string]string{}, // TODO: include configuration flags
+		func(f http.HandlerFunc) http.HandlerFunc { return f },
+		func() v1.TSDBAdmin { return nil }, // Only needed for admin APIs.
+		false,                              // Disable admin APIs.
+		a.logger,
+		querier.DummyRulesRetriever{},
+		0, 0, 0, // Remote read samples and concurrency limit.
+		regexp.MustCompile(".*"),
+		func() (v1.RuntimeInfo, error) { return v1.RuntimeInfo{}, errors.New("not implemented") },
+		&v1.PrometheusVersion{},
+	)
+
+	promRouter := route.New().WithPrefix(a.cfg.ServerPrefix + a.cfg.PrometheusHTTPPrefix + "/api/v1")
```

Comment:
Why prefixing with `ServerPrefix`? Looks something we're just doing here.

Reply:
This is a bugfix. Currently the `server.path-prefix` flag will break all requests routed to the promRouter. I have wrote up why in this issue: https://github.com/cortexproject/cortex/issues/2411.